### PR TITLE
Refactor xDS Cache

### LINKED ISSUE #102: Refactor CacheFactory and CacheFlush
**Why is this needed**:

```
type CacheFactory interface {
	StatusFlush(status core_v2.ApiStatus) int
	StatusDelete(status core_v2.ApiStatus)
	StatusReset(old, new core_v2.ApiStatus)
}
```

CacheFactory is the interface to update and delete xds resources that received from xds control plane. 

But currently the cache flushing procedure is quite hard to understand,  And there are other caveats:

1. It does not support delta cache flush, which is useful for delta xds.
1. the underlying implementor of CacheFactory stores redundant resources

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**What type of PR is this?**

/kind enhancement

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #102

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
@nlgwcy Open up for early feedback

**Comment by hzxuzhonghu** [comment]
cc @YaoZengzeng

**Comment by hzxuzhonghu** [comment]
ping @nlgwcy @YaoZengzeng

File: pkg/controller/envoy/ads_event.go [hunk 1]
```diff
@@ -227,25 +237,25 @@ func (svc *ServiceEvent) handleRdsResponse(rsp *service_discovery_v3.DiscoveryRe
...
+		newHash := hash.Sum64String(resource.String())
+		if newHash != svc.DynamicLoader.RouteCache.GetRdsHash(routeConfiguration.GetName()) {
+			svc.DynamicLoader.RouteCache.SetRdsHash(routeConfiguration.GetName(), newHash)
 			log.Debugf("[CreateApiRouteByRds] update %s", routeConfiguration.GetName())
 		} else {
 			apiStatus = core_v2.ApiStatus_UNCHANGED
 		}
-
 		svc.DynamicLoader.CreateApiRouteByRds(apiStatus, routeConfiguration)
 	}
```
Comment (YaoZengzeng) [inline pkg/controller/envoy/ads_event.go:256]: Doesn't rds need to "diff then remove" ?

File: pkg/controller/envoy/ads_event.go [hunk 2]
```diff
@@ -131,16 +132,18 @@ func (svc *ServiceEvent) handleCdsResponse(rsp *service_discovery_v3.DiscoveryRe
...
 	)
 
+	current := sets.New[string]()
 	for _, resource := range rsp.GetResources() {
 		if err = anypb.UnmarshalTo(resource, cluster, proto.UnmarshalOptions{}); err != nil {
 			continue
 		}
+		current.Insert(cluster.GetName())
 		// compare part[0] CDS now
 		// Cluster_EDS need compare tow parts, compare part[1] EDS in EDS handler
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pkg/controller/envoy/ads_event.go:142]: ok

File: pkg/controller/envoy/ads_event.go [hunk 3]
```diff
@@ -192,32 +200,34 @@ func (svc *ServiceEvent) handleLdsResponse(rsp *service_discovery_v3.DiscoveryRe
...
-
 		apiStatus := core_v2.ApiStatus_UPDATE
-		newLdsResource := resource.String()
-		if newLdsResource != svc.DynamicLoader.ListenerCache.GetLdsResource(listener.GetName()) {
-			svc.DynamicLoader.ListenerCache.SetLdsResource(listener.GetName(), newLdsResource)
+		current.Insert(listener.GetName())
+		newHash := hash.Sum64String(resource.String())
+		if newHash != svc.DynamicLoader.ListenerCache.GetLdsHash(listener.GetName()) {
+			svc.DynamicLoader.ListenerCache.AddOrUpdateLdsHash(listener.GetName(), newHash)
 			log.Debugf("[CreateApiListenerByLds]update %s", listener.GetName())
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pkg/controller/envoy/ads_event.go:213]: Not new introduced, will fix though

File: pkg/controller/envoy/ads_event.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/controller/envoy/ads_event.go:256]: Missed that

**Pull request merged**